plugins/keep-markup/tests/test.js

```diff
@@ -0,0 +1,92 @@
+const expect = require('chai').expect;
+const jsdom = require('jsdom')
+const { JSDOM } = jsdom
+
+require('../../../prism')
+// fake DOM
+global.self = {}
+global.self.Prism = Prism
+global.document = {}
+document.createRange = function () {
+}
+global.self.document = document
+
+require('../prism-keep-markup')
+
+describe('Prism Keep Markup Plugin', function () {
+
+	function execute (code) {
+		const start = [];
+		const end = [];
+		const nodes = [];
+		document.createRange = function () {
+			return {
+				setStart: function (node, offset) {
+					start.push({ node, offset })
+				},
+				setEnd: function (node, offset) {
+					end.push({ node, offset })
+				},
+				extractContents: function () {
+					return new JSDOM('').window.document.createTextNode('')
+				},
+				insertNode: function (node) {
+					nodes.push(node)
+				},
+				detach: function () {
+				}
+			}
+		}
+		const beforeHighlight = Prism.hooks.all['before-highlight'][0]
+		const afterHighlight = Prism.hooks.all['after-highlight'][0]
+		const env = {
+			element: new JSDOM(code).window.document.getElementsByTagName('code')[0],
+			language: "javascript"
+		}
+		beforeHighlight(env)
+		afterHighlight(env)
+		return { start, end, nodes }
+	}
+
+	it('should keep <span> markup', function () {
+		const result = execute(`<code class="language-javascript">x<span>a</span>y</code>`)
+		expect(result.start.length).to.equal(1)
+		expect(result.end.length).to.equal(1)
+		expect(result.nodes.length).to.equal(1)
+		expect(result.nodes[0].nodeName).to.equal('SPAN')
+	})
+	it('should preserve markup order', function () {
+		const result = execute(`<code class="language-javascript">x<a></a><b></b>y</code>`)
+		expect(result.start.length).to.equal(2)
+		expect(result.start[0].offset).to.equal(0)
+		expect(result.start[0].node.textContent).to.equal('y')
+		expect(result.start[1].offset).to.equal(0)
+		expect(result.start[1].node.textContent).to.equal('y')
+		expect(result.end.length).to.equal(2)
+		expect(result.end[0].offset).to.equal(0)
+		expect(result.end[0].node.textContent).to.equal('y')
+		expect(result.end[1].offset).to.equal(0)
+		expect(result.end[1].node.textContent).to.equal('y')
+		expect(result.nodes.length).to.equal(2)
+		expect(result.nodes[0].nodeName).to.equal('A')
+		expect(result.nodes[1].nodeName).to.equal('B')
+	})
+	it('should keep last <span> markup', function () {
+		const result = execute(`<code class="language-javascript">xy<span>a</span></code>`)
+		expect(result.start.length).to.equal(1)
+		expect(result.end.length).to.equal(1)
+		expect(result.nodes.length).to.equal(1)
+		expect(result.nodes[0].nodeName).to.equal('SPAN')
+	})
```

Comment:
I almost think snapshot tests would be better here than all these assertions. The other thing I've done is something like:

```js
expect(result.innerHtml).to.equal('<some>html</some>`);
```

If we didn't want to add a new dep & auto-generate them.

Reply:
> I almost think snapshot tests would be better here than all these assertions. The other thing I've done is something like:

You tell me 😉 
Do you know a good snapshot testing library ? In my opinion we should do what you suggest:

```js
expect(result.innerHtml).to.equal('<some>html</some>');
 ```

Keep it simple.

Reply:
I really like this simple approach.
We can always add a more elaborate testing scheme later on.

Reply:
Great, no additional lib needed then. Alternatively: https://github.com/suchipi/chai-jest-snapshot

Reply:
> Alternatively: https://github.com/suchipi/chai-jest-snapshot

Maybe I don't get the library but don't they separate the expected result from the actual input? Wouldn't it be better to have actual and expected in the same file in our case?

Reply:
I just get back to the code and it won't work because we are using a dummy implementation of `createRange` (since `jsdom` doesn't implement it).
So `innerHTML` will be empty/undefined.